CMakeLists.txt

```diff
@@ -5,18 +5,44 @@ cmake_minimum_required(VERSION 3.11)
 set(CCF_DIR ${CMAKE_CURRENT_SOURCE_DIR})
 include(${CCF_DIR}/cmake/preproject.cmake)
 
+# Deduce project version from git environment
+if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/.git)
+  FIND_PACKAGE(Git)
+  execute_process(
+    COMMAND git describe --tags
+    OUTPUT_VARIABLE "CCF_VERSION"
+    OUTPUT_STRIP_TRAILING_WHITESPACE
+  )
+  execute_process(
+    COMMAND bash "-c" "git describe --tags --abbrev=0 | cut -d'-' -f1 | tr -d v"
+    OUTPUT_VARIABLE "CCF_RELEASE_VERSION" OUTPUT_STRIP_TRAILING_WHITESPACE
+  )
+  # Check that git release tag follows semver
+  execute_process(
+    COMMAND
+      bash "-c"
+      "[[ ${CCF_RELEASE_VERSION} =~ ^([[:digit:]])+(\.([[:digit:]])+)*(-.*)?$ ]]"
+    RESULT_VARIABLE "TAG_IS_SEMVER"
+  )
+  if(NOT ${TAG_IS_SEMVER} STREQUAL "0")
+    message(
+      FATAL_ERROR "git tag \"${CCF_RELEASE_VERSION}\" does not follow semver"
+    )
+  endif()
+else()
+  set(CCF_VERSION "unknown")
+  set(CCF_RELEASE_VERSION "0.0")
```

Comment:
If we're not in git (e.g. release tarball), what should we do? We could do what we used to do (manual update) but this will be out-of-sync sooner than later or agree on a nice default that will work for all versions.

Reply:
For releases, we could produce a file, and look up that file. For zip downloads that aren't checkouts, perhaps use the name of the directory?

Reply:
I'm leaning towards keeping this up-to-date manually, with a release check/build flag that the 2 versions match to prevent it getting out-of-sync. So in a normal in-Git build we use the Git version. If we can't find Git we use `CCF_EXPLICIT_VERSION`. If we have `-DCHECK_EXPLICIT_VERSION_MATCH=ON` we compare the Git-retrieved version and `CCF_EXPLICIT_VERSION` and fail if they don't match. We only set the latter for release builds.

This seems best of both worlds - we get `version-hash` for nearly all builds, but non-Git release tarballs are still built with the correct version.

Reply:
As discussed, let's go for parent directory name deduction (e.g. `CCF-0.11` for release tarball).